File: plugins/skill-intake/scripts/convert_md_to_json.js
#!/usr/bin/env node
'use strict';

const fs = require('node:fs');
const path = require('node:path');

const AXIS_HEADINGS = {
  output_destination: ['出力先', 'Output Destination'],
  info_source: ['情報源', 'Information Source'],
  share_target: ['共有相手', 'Sharing Target'],
  true_problem: ['真の課題', 'True Problem'],
  knowledge_assets: ['ナレッジ資産', 'Knowledge Assets'],
};

function parseSections(md) {
  const lines = md.split(/\r?\n/);
  const sections = {};
  let current = null;
  let buf = [];
  for (const line of lines) {
    const m = line.match(/^#{1,4}\s+(.*?)\s*$/);
    if (m) {
      if (current) sections[current] = buf.join('\n').trim();
      current = m[1];
      buf = [];
    } else if (current) {
      buf.push(line);
    }
  }
  if (current) sections[current] = buf.join('\n').trim();
  return sections;
}

function pickAxis(sections, keys) {
  for (const k of Object.keys(sections)) {
    if (keys.some((needle) => k.includes(needle))) return sections[k];
  }
  return '';
}

function extractFrontmatter(md) {
  const m = md.match(/^---\n([\s\S]*?)\n---\n/);
  if (!m) return { meta: {}, body: md };
  const meta = {};
  for (const line of m[1].split(/\n/)) {
    const kv = line.match(/^(\w[\w_-]*)\s*:\s*(.*)$/);
    if (kv) meta[kv[1]] = kv[2].trim().replace(/^['"]|['"]$/g, '');
  }
  return { meta, body: md.slice(m[0].length) };
}

function convert(md) {
  const { meta, body } = extractFrontmatter(md);
  const sections = parseSections(body);
  const five_axes = {};
  for (const k of Object.keys(AXIS_HEADINGS)) five_axes[k] = pickAxis(sections, AXIS_HEADINGS[k]);
  const out = {
    skill_name_hint: meta.skill_name_hint || meta.name || '',
    pattern: meta.pattern || 'other',
    user_profile: sections['User Profile'] || sections['利用者プロファイル'] || '',
    '5_axes': five_axes,
    sections,
    open_questions: [],
    integrations: meta.integrations ? meta.integrations.split(/\s*,\s*/) : [],
    raw_meta: meta,
  };
  return out;
}

function main(argv) {
  const file = argv[2];
  const outFile = argv[3];
  if (!file) { process.stderr.write('usage: convert_md_to_json.js <intake.md> [intake.json]\n'); return 2; }
  let md;
  try { md = fs.readFileSync(path.resolve(file), 'utf8'); }
  catch (e) { process.stderr.write(`input error: ${e.message}\n`); return 2; }
  const json = convert(md);
  const text = JSON.stringify(json, null, 2);
  if (outFile) fs.writeFileSync(path.resolve(outFile), text + '\n');
  else process.stdout.write(text + '\n');
  return 0;
}

module.exports = { convert, parseSections };

if (require.main === module) {
  process.exit(main(process.argv));
}

File: plugins/skill-intake/scripts/convert_md_to_json.py
#!/usr/bin/env python3
"""Convert intake.md (front-matter + sections) to intake.json."""
from __future__ import annotations

import json
import re
import sys
from pathlib import Path
from typing import Any

AXIS_HEADINGS: dict[str, list[str]] = {
    'output_target': ['出力先', 'Output Destination'],
    'info_source': ['情報源', 'Information Source'],
    'share_target': ['共有相手', 'Sharing Target'],
    'true_problem': ['真の課題', 'True Problem'],
    'knowledge_assets': ['ナレッジ資産', 'Knowledge Assets'],
}

_HEADING = re.compile(r'^#{1,4}\s+(.*?)\s*$')
_FRONTMATTER = re.compile(r'^---\n([\s\S]*?)\n---\n')
_FM_KV = re.compile(r'^(\w[\w_-]*)\s*:\s*(.*)$')
_INTENT_BLOCK = re.compile(
    r'###\s+入力→出力 intent contract\s*```json\s*([\s\S]*?)\s*```',
    re.M,
)


def parse_sections(md: str) -> dict[str, str]:
    sections: dict[str, str] = {}
    current: str | None = None
    buf: list[str] = []
    for line in md.splitlines():
        m = _HEADING.match(line)
        if m:
            if current is not None:
                sections[current] = '\n'.join(buf).strip()
            current = m.group(1)
            buf = []
        elif current is not None:
            buf.append(line)
    if current is not None:
        sections[current] = '\n'.join(buf).strip()
    return sections


def pick_axis(sections: dict[str, str], keys: list[str]) -> str:
    for k in sections:
        if any(needle in k for needle in keys):
            return sections[k]
    return ''


def parse_five_axes_table(body: str) -> dict[str, str]:
    """Extract §6 5-axis summary table from intake-final-template output."""
    out: dict[str, str] = {}
    axis_by_label = {
        '出力先': 'output_target',
        '情報源': 'info_source',
        '共有相手': 'share_target',
        '真の課題': 'true_problem',
        'ナレッジ資産': 'knowledge_assets',
    }
    for line in body.splitlines():
        if not line.startswith('|'):
            continue
        cells = [c.strip() for c in line.strip().strip('|').split('|')]
        if len(cells) < 3 or cells[0] in ('軸', '---'):
            continue
        label = re.sub(r'[*`]', '', cells[0]).replace('（MUST）', '').strip()
        key = axis_by_label.get(label)
        if key:
            out[key] = cells[1].strip()
    return out


def parse_intent_contract(body: str) -> dict[str, Any] | None:
    """Extract fenced intent_contract JSON from §6, if present."""
    m = _INTENT_BLOCK.search(body)
    if not m:
        return None
    try:
        value = json.loads(m.group(1))
    except json.JSONDecodeError:
        return None
    return value if isinstance(value, dict) else None


def normalize_pattern(raw: str) -> str:
    value = (raw or '').strip().upper()
    if value in {'A', 'B', 'C', 'D', 'E'}:
        return value
    m = re.search(r'パターン:\s*([A-E])', raw or '')
    if m:
        return m.group(1)
    return 'E'


def pick_user_profile(sections: dict[str, str]) -> str:
    for key, value in sections.items():
        if key in ('User Profile', '利用者プロファイル') or 'user-profiler' in key or 'ユーザー像' in key:
            return value
    return ''


def pick_skill_name(meta: dict[str, str], body: str) -> str:
    if meta.get('skill_name_hint') or meta.get('name'):
        return meta.get('skill_name_hint') or meta.get('name') or ''
    m = re.search(r'^#\s+(.+?)(?:\s+—|\s+-|\n)', body, re.M)
    if m:
        raw = m.group(1).strip()
        slug = re.sub(r'[^a-z0-9-]+', '-', raw.lower()).strip('-')
        return slug or raw
    return 'intake-final'


def extract_frontmatter(md: str) -> tuple[dict[str, str], str]:
    m = _FRONTMATTER.match(md)
    if not m:
        return {}, md
    meta: dict[str, str] = {}
    for line in m.group(1).split('\n'):
        kv = _FM_KV.match(line)
        if kv:
            meta[kv.group(1)] = kv.group(2).strip().strip("'\"")
    return meta, md[m.end():]


def convert(md: str) -> dict[str, Any]:
    meta, body = extract_frontmatter(md)
    sections = parse_sections(body)
    five_axes = parse_five_axes_table(body)
    intent_contract = parse_intent_contract(body)
    if not five_axes:
        five_axes = {k: pick_axis(sections, v) for k, v in AXIS_HEADINGS.items()}
    if intent_contract:
        five_axes['intent_contract'] = intent_contract
    integrations_raw = meta.get('integrations', '')
    integrations = [s.strip() for s in re.split(r'\s*,\s*', integrations_raw)] if integrations_raw else []
    pattern = normalize_pattern(meta.get('pattern') or sections.get('skill') or body[:500])
    return {
        'skill_name_hint': pick_skill_name(meta, body),
        'pattern': pattern,
        'user_profile': pick_user_profile(sections),
        '5_axes': five_axes,
        'sections': sections,
        'open_questions': [],
        'integrations': integrations,
        'raw_meta': meta,
        'validation': {
            'render': 'PASS',
            'quality_gate': 'PASS',
            'cross_check': 'PASS',
            'failures': [],
        },
    }


def main(argv: list[str]) -> int:
    if len(argv) < 1:
        sys.stderr.write('usage: convert_md_to_json.py <intake.md> [intake.json]\n')
        return 2
    file = argv[0]
    out_file = argv[1] if len(argv) > 1 else None
    try:
        md = Path(file).resolve().read_text(encoding='utf-8')
    except Exception as e:
        sys.stderr.write(f'input error: {e}\n')
        return 2
    j = convert(md)
    text = json.dumps(j, ensure_ascii=False, indent=2)
    if out_file:
        Path(out_file).resolve().write_text(text + '\n', encoding='utf-8')
    else:
        sys.stdout.write(text + '\n')
    return 0


if __name__ == '__main__':
    sys.exit(main(sys.argv[1:]))
